packages/ui-core/src/components/Select/Select.tsx

```diff
@@ -1,481 +1,439 @@
+
+        if (type === Types.COMBOBOX) {
+            this.inputNode.value = title;
         }
 
-        this.setState({ isOpen: true });
+        onSelect && onSelect(e, item);
     }
 
-    handleChangeSearch = (e: React.SyntheticEvent<HTMLInputElement>): void => {
+    handleHoverItem = (index: number) => (e: React.SyntheticEvent<EventTarget>): void => {
+        e.preventDefault();
+
+        this.setState({ activeIndex: index });
+    }
+
+    handleClickCombobox = (e: React.FormEvent<EventTarget>): void => {
+        const { isOpened, filteredItems } = this.state;
+
+        e.stopPropagation();
+
         if (!(e.target instanceof HTMLInputElement)) {
             return;
         }
 
-        const { onChangeSearch } = this.props;
-        onChangeSearch && onChangeSearch(e.target.value);
+        if (!isOpened && filteredItems) {
+            e.target.select();
+        }
 
-        this.setState({
-            activeIndex: 0,
-            currentIndex: -1,
-            isOpen: true,
-        });
+        this.setState((state) => ({ isOpened: !state.isOpened }));
     }
 
-    handleFocusSearch = (e: React.SyntheticEvent<HTMLInputElement>): void => {
-        if (!(e.target instanceof HTMLInputElement)) {
-            return;
-        }
+    handleChangeCombobox = (e: React.ChangeEvent<HTMLInputElement>): void => {
+        const { items } = this.props;
+        const filterValue = e.target.value;
+
+        const filteredItems = items.filter(({ title }) => {
+            if (filterValue.length <= title.length) {
+                return RegExp(filterValue, 'ig').test(title);
+            }
 
-        const { onFocusSearch } = this.props;
+            return false;
+        });
 
-        onFocusSearch && onFocusSearch(e.target.value);
+        this.debouncedComboboxChange(filteredItems, filterValue);
     }
 
     handleKeyDown = (e: React.KeyboardEvent<HTMLDivElement>): boolean => {
-        if (!this.props.keyNavigation) {
-            return false;
-        }
+        const { activeIndex, isOpened } = this.state;
+        const { items } = this.props;
 
-        const { activeIndex } = this.state;
+        if (items.length === 0) {
+            return true;
+        }
 
-        // key arrow down
-        if (e.keyCode === 40 && activeIndex < this.props.items.length - 1) {
+        if (e.key === 'ArrowDown' && activeIndex < items.length - 1) {
             this.setState({ activeIndex: activeIndex + 1 }, () => {
                 this.scrollList(this.state.activeIndex);
             });
+
             e.preventDefault();
+
             return false;
         }
-        // key arrow up
-        if (e.keyCode === 38 && activeIndex > 0) {
-            this.setState({ activeIndex: activeIndex - 1 }, () => {
+        if (e.key === 'ArrowUp' && activeIndex > 0) {
+            this.setState((prevState) => ({ activeIndex: prevState.activeIndex - 1 }), () => {
                 this.scrollList(this.state.activeIndex);
```

Comment:
По идее, 
```suggestion
                this.scrollList(activeIndex);
```
Ибо это одно и тоже.

Reply:
Нет, не одно и то же. Внутри коллбека в сетстейте по-разному определяется текущий стейт. Видимо это как-то связано с константой, в которой я деструктурирую стэйт. Если обращаюсь к activeIndex, который лежит в деструктуризации, то там он еще не измененный. Из-за этого и происходит косяк с перемещением по списку посредством клавиш, происходит запаздывание на один элемент.

Reply:
М, да, действительно, туплю. В callback будет обновленный стейт, а activeIndex из деструкторизации остается прежней из-за замыкания.